velox/common/caching/SsdCache.h

```diff
@@ -0,0 +1,355 @@
+  }
+
+  SsdRun(const SsdRun& other) = default;
+  SsdRun(SsdRun&& other) = default;
+
+  void operator=(const SsdRun& other) {
+    bits_ = other.bits_;
+  }
+  void operator=(SsdRun&& other) {
+    bits_ = other.bits_;
+  }
+
+  uint64_t offset() const {
+    return (bits_ >> kSizeBits);
+  }
+
+  uint32_t size() const {
+    return (bits_ & ((1 << kSizeBits) - 1)) + 1;
+  }
+
+ private:
+  uint64_t bits_;
+};
+
+// Key for SsdFile lookup. The key is the file number in storage and
+// the offset in the file. SSD cache sizes align to the RAM cache
+// quantized sizes for cached streams from the original file.
+struct SsdKey {
+  StringIdLease file;
+  uint64_t offset;
+
+  bool operator==(const SsdKey& other) const {
+    return offset == other.offset && file.id() == other.file.id();
+  }
+};
+
+} // namespace facebook::velox::cache
+namespace std {
+template <>
+struct hash<::facebook::velox::cache::SsdKey> {
+  size_t operator()(const ::facebook::velox::cache::SsdKey& key) const {
+    return facebook::velox::bits::hashMix(key.file.id(), key.offset);
+  }
+};
+
+} // namespace std
+
+namespace facebook::velox::cache {
+
+// Represents an SsdFile entry that is planned for load or being
+// loaded. This is destroyed after load. Destruction decrements the
+// pin count of the corresponding region of 'file_'. While there are
+// pins, the region cannot be evicted.
+class SsdPin {
+ public:
+  SsdPin() : file_(nullptr) {}
+
+  // Constructs a pin referencing 'run' in 'file'. The region must be
+  // pinned before constructing the pin.
+  SsdPin(SsdFile& file, SsdRun run);
+
+  SsdPin(const SsdPin& other) = delete;
+
+  SsdPin(SsdPin&& other) {
+    run_ = other.run_;
+    file_ = other.file_;
+    other.file_ = nullptr;
+  }
+
+  ~SsdPin();
+
+  void operator=(SsdPin&&);
+
+  bool empty() const {
+    return file_ == nullptr;
+  }
+  SsdFile* file() const {
+    return file_;
+  }
+
+  SsdRun run() const {
+    return run_;
+  }
+
+ private:
+  SsdFile* file_;
+  SsdRun run_;
+};
+
+// Metrics for SSD cache. Maintained by SsdFile and aggregated by SsdCache.
+struct SsdCacheStats {
+  uint64_t entriesWritten{0};
+  uint64_t bytesWritten{0};
+  uint64_t entriesRead{0};
+  uint64_t bytesRead{0};
+  uint64_t entriesCached{0};
+  uint64_t bytesCached{0};
+};
+
+// A shard of SsdCache. Corresponds to one file on SSD.  The data
+// backed by each SsdFile is selected on a hash of the storage file
+// number of the cached data. Each file consists of an integer number
+// of 64MB regions. Each region has a pin count and an read
+// count. Cache replacement takes place region by region, preferring
+// regions with a smaller read count. Entries do not span
+// regions. Otherwise entries are consecutive byte ranges inside
+// their region.
+class SsdFile {
+ public:
+  static constexpr uint64_t kMaxSize = 1UL << 36; // 64G
+  static constexpr uint64_t kRegionSize = 1 << 26; // 64MB
+  static constexpr int32_t kNumRegions = kMaxSize / kRegionSize;
```

Comment:
Looks like `kNumRegions` is not used and therefore `kMaxSize` is not used anywhere.